### ISSUE TITLE
Invalid signer error when using factory instance connect method

### ISSUE THREAD

**Issue by mdnorman** [dethcrypto/TypeChain#487]
Problem:

When using the factory _instance_ `connect` method, ethers gives back an error that simply says `invalid signer`. This is due to a mismatch of the constructors on the generated factory subclasses and the `ContractFactory` base class.

Versions:

* `"@typechain/ethers-v5": "7.1.2"`
* `"ethers": "5.4.7"`

Simple repro:

NOTE: Any set of typechain-generated classes will do this, so it's not specific to my contracts, but I'm using a simple contract with no constructor parameters to demonstrate the issue.

```
const signer = getValidSigner(); // assume that this returns a valid signer eg a wallet

// the following works great:
const accessControlFacet = await new AccessControlFacet__factory(signer).deploy();
console.log('accessControlFacet address:', accessControlFacet.address);

// the following fails on the first line at `connect(signer)`:
const accessControlFacetFactory = new AccessControlFacet__factory().connect(signer);
const accessControlFacet2 = await accessControlFacetFactory.deploy();
console.log('accessControlFacet2 address:', accessControlFacet2.address);
```

Root cause:

The root cause is that the constructor for `AccessControlFacet__factory` takes in just the signer (as can be seen above), but the `ContractFactory` class constructor takes in the ABI, bytecode, and the signer as the 3rd parameter. Following the trail of the generated `connect` function, I see this:

```
AccessControlFacet_factory.ts:
  ...
export class AccessControlFacet__factory extends ContractFactory {
...
  connect(signer: Signer): AccessControlFacet__factory {
    return super.connect(signer) as AccessControlFacet__factory;
  }
  ...
}

@ethersproject/contracts/src/index.ts:
...
export class ContractFactory {
...
    constructor(contractInterface: ContractInterface, bytecode: BytesLike | { object: string }, signer?: Signer) {
...
        // If we have a signer, make sure it is valid
        if (signer && !Signer.isSigner(signer)) {
            logger.throwArgumentError("invalid signer", "signer", signer);
        }
...
  }
...
    connect(signer: Signer) {
        return new (<{ new(...args: any[]): ContractFactory }>(this.constructor))(this.interface, this.bytecode, signer);
    }
...
}
```

The failing line is in the constructor, but the culprit is in `connect`.

Notice that the first parameter to `this.constructor` is the interface, and indeed when I debug into the `connect` function, the `signer` is good to go. However, it passes it as the third parameter to the constructor, but the `this.constructor` is actually `AccessControlFacet__factory`'s constructor, which only takes a single parameter. This means the `signer` field in the `AccessControlFacet__factory` constructor is actually interface, as defined in `this.interface`.

One could argue that this is an ethers project problem, but the assumption by that project is that the constructor always has 3 parameters for the subclasses (maybe rightly so?). To get around this in the typechain-generated classes, it seems like the `connect` method can't be used and instead the constructor should be used directly, since that's really all it needs to do anyway.

I hope this helps. Happy to provide more details if needed.

**Comment by zemse** [comment]
```ts
// the following fails on the first line at `connect(signer)`:
const accessControlFacetFactory = new AccessControlFacet__factory().connect(signer);
```

if you change the above to the following does that fix the error ?

```ts
const accessControlFacetFactory = new AccessControlFacet__factory(signer);
```

**Comment by mdnorman** [comment]
Yep, that works. The problem is that I need to use the factory as a generic factory that I've passed into another function.

**Comment by zemse** [comment]
> The problem is that I need to use the factory as a generic factory that I've passed into another function.

Do you mean you want to construct a factory without a signer (then pass to another function and attach signer there) or could you elaborate?

**Comment by hasparus** [comment]
Could we rewrite `AccessControlFacet__factory::connect` to use the constructor instead of calling `super.connect`?

#### After

```ts
  connect(signer: Signer): Auction__factory {
    return new Auction__factory(signer);
  }
```

#### Currently

```ts
  connect(signer: Signer): Auction__factory {
    return super.connect(signer) as Auction__factory;
  }
```

**Comment by mdnorman** [comment]
I think either of those solutions would work for my use cases. The constructor solution makes it so that subclasses of AccessControlFacet__factory would still work, but I'm not sure that's necessarily needed.

**Comment by mdnorman** [comment]
> > The problem is that I need to use the factory as a generic factory that I've passed into another function.
> 
> Do you mean you want to construct a factory without a signer (then pass to another function and attach signer there) or could you elaborate?

Yes, that's exactly correct. Imagine I have this code:

```
const myFunc = (factory: ContractFactory) => {
  const signer = getSigner();
  const factoryWithSigner = factory.connect(signer);
  return factoryWithSigner.deploy();
}
```

Simplistic example, but you get the drift.
